### ISSUE TITLE
Better documentation on _mask and _key_padding_mask in transformers modules

### ISSUE THREAD

**Issue by kirk86** [pytorch/pytorch#120488]
### 📚 The doc issue

The [docs](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html) here do not clearly indicate the shapes of the masks and their equivalent types that the transformer modules are expecting.

The only place where one can find shapes and types of expected tensors are only in the [source code docs](https://github.com/pytorch/pytorch/blob/54bac042e711f08e909dda939f70d6f272dbe8ce/torch/csrc/api/include/torch/nn/modules/transformer.h#L45) atm.

### Suggest a potential alternative/fix

A couple of examples of different use cases would really help users understand what is expected what the actual effect is of each mask. 

Would the transformer module operate on both of of them?

Also, a clarification on why one uses a boolean mask for `_key_padding_mask` with `True` values indicating there's padding there and `False` there's no padding which is actually the opposite of other frameworks.

For instance, here's an excerpt from hugging face explaining the same thing but they use 1s, i.e., `True` indicating where the model needs to attend to and 0s `False` where there's padding.

```
Attention masks
Attention masks are tensors with the exact same shape as the input IDs tensor, filled with 0s and 1s.
1s indicate the corresponding tokens should be attended to, 
and 0s indicate the corresponding tokens should not be attended to 
(i.e., they should be ignored by the attention layers of the model).
```

cc @svekars @brycebortree @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki
